src/Microsoft.OData.Core/UriParser/Binders/CountSegmentBinder.cs

```diff
@@ -0,0 +1,72 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="CountSegmentBinder.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.UriParser
+{
+    using ODataErrorStrings = Microsoft.OData.Strings;
+
+    /// <summary>
+    /// Class that knows how to bind a Count segment token.
+    /// </summary>
+    internal sealed class CountSegmentBinder
+    {
+        /// <summary>
+        /// Method to use to visit the token tree and bind the tokens recursively.
+        /// </summary>
+        private readonly MetadataBinder.QueryTokenVisitor bindMethod;
+
+        /// <summary>
+        /// State to use for binding.
+        /// </summary>
+        private readonly BindingState state;
+
+        /// <summary>
+        /// Constructs a CountSegmentBinder object.
+        /// </summary>
+        /// <param name="bindMethod">Method to use to visit the token tree and bind the tokens recursively.</param>
+        /// <param name="state">State of the metadata binding.</param>
+        internal CountSegmentBinder(MetadataBinder.QueryTokenVisitor bindMethod, BindingState state)
+        {
+            this.bindMethod = bindMethod;
+            this.state = state;
+        }
+
+        /// <summary>
+        /// Binds an Count segment token.
+        /// </summary>
+        /// <param name="countSegmentToken">The Count segment token to bind.</param>
+        /// <param name="state">State of the metadata binding.</param>
+        /// <returns>The bound Count segment token.</returns>
+        internal QueryNode BindCountSegment(CountSegmentToken countSegmentToken)
+        {
+            ExceptionUtils.CheckArgumentNotNull(countSegmentToken, "countSegmentToken");
+
+            QueryNode source = this.bindMethod(countSegmentToken.NextToken);
+            CollectionNode node = source as CollectionNode;

```

Comment:
>node [](http://example.com/codeflow?start=27&length=4)

any chance this could be null?  i.e., that bindMethod could return null, or something other than a CollectionNode?

Reply:
I have added the code to throw an exception whenever the node is null.

Reply:
This is an internal method, so if this would only be null i we had a logic issue in our code we could make this an assert, rather than a null check in production code

Reply:
Since this occurs when user tries parsing `NonCollectionProperty/$count`, we need to return the exception to the user.